### PR TITLE
Cleanup typechecking

### LINKED ISSUE #673: alternative formats for `ty` output
... like we used to have before #500.

I guess we need a new set of protobuf messages for "get me the typing error in format X" unless there's a wire format for `ty` exceptions we can use to send the error, then format it at the other end, but I don't think there is.

At the same time, we should probably get rid of the pool bs in type checking.

### PR DESCRIPTION
**Pull request by samuelcolvin**
fix #673.

---
## Summary by cubic
Switched type checking to a per-session `TypeChecker` and moved diagnostic rendering into workers with configurable format/color. Also fixes a JS format-encoding edge case, surfaces reset errors as `FatalError`, corrects REPL prompt styling when stdout is redirected, and adds reuse-guaranteed tests that pin cross-session isolation across worker resets.

- **Refactors**
  - Replaced `type_check(...)` with `TypeChecker::run(...)`; added `TypeChecker::reset()` and used it on worker configure/recycle.
  - Introduced `TypeCheckingConfig` and `TypeCheckingFormat` in `monty-types`; the config travels on `monty-proto::Configure` (new enum and color flag) and is persisted in dumps.
  - Removed the pooled DB and implemented scoped file tracking/cleanup inside `TypeChecker` (prevents repeated-path entries growing per feed).
  - Workers (including WASM) render diagnostics; CLI gains `--type-check-format`/`--type-check-color` and now uses `anstream`/`anstyle` for styling (respects NO_COLOR/CLICOLOR and enables Windows VT; REPL and checker share the same decision).
  - JS/Python surface the format via checkout options: JS `typeCheckFormat`/`typeCheckColor`, Python `type_check_format`/`type_check_color`; `monty-js` removes the native `MontyTypingError` wrapper in favor of the TS class.
  - Python: consolidated format names into a single `TypeCheckFormat` type alias in `pydantic_monty`; stubs import it, and `type_check_format` now defaults to `None` (pyo3 `Option`) rather than a stubbed ellipsis.
  - Fixes: wasm transport now validates `typeCheckFormat` as an own property (avoids prototype-chain lookups), worker `Reset` emits a `FatalError` instead of crashing if `TypeChecker::reset()` fails, and the REPL prompt styles from stdout so redirection doesn’t leak escapes.
  - Tests: added reuse-guaranteed isolation tests for subprocess, pool, and Python bindings; assert same worker reuse, probe module/stub cleanup across `Reset`, and cover script-name path shapes.

- **Migration**
  - Replace `type_check(&SourceFile::new(code, path), stubs)` with `TypeChecker::default().run(&SourceFile::new(code, path), stubs, TypeCheckingConfig::default())`; call `TypeChecker::reset()` between sessions if reusing an instance.
  - Choose the rendering at checkout: JS `typeCheckFormat`/`typeCheckColor`, Python `type_check_format`/`type_check_color`, or CLI `--type-check-format`/`--type-check-color`.

<sup>Written for commit a13bafea8b89f1cf1f99f234cd7b46874f187d55. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/pydantic/monty/pull/674?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: crates/monty-type-checking/src/type_check.rs [hunk 1]
```diff
@@ -35,87 +30,139 @@ impl<'a> SourceFile<'a> {
...
+    /// * `Ok(None)` - If there are no typing errors.
+    /// * `Ok(Some(string))` - If there are typing errors.
+    /// * `Err(String)` - If there was an unexpected/internal error during type checking.
+    pub fn run<'a>(
+        &'a mut self,
+        python_source: &SourceFile<'_>,
+        stubs_file: Option<&SourceFile<'_>>,
+        config: TypeCheckingConfig,
+    ) -> Result<Option<TypeCheckingDiagnostics<'a>>, String> {
+        let src_root = SystemPathBuf::from(SRC_ROOT);
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply crates/monty-type-checking/src/type_check.rs:57]: no, we want to run repeated type checks without deleting existing files.

File: crates/monty-runtime/src/main.rs [hunk 2]
```diff
@@ -189,7 +194,11 @@ fn main() -> ExitCode {
 
 /// Dispatches a parsed standalone CLI invocation.
 fn run_cli(cli: Cli) -> ExitCode {
-    let type_check_enabled = cli.type_check;
+    // `Some` enables the check and carries how its diagnostics are rendered.
+    let type_check = cli.type_check.then(|| TypeCheckingConfig {
+        format: cli.type_check_format.unwrap_or_default(),
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply crates/monty-runtime/src/main.rs:219]: fine for now.

File: crates/monty-python/python/pydantic_monty/_monty.pyi [hunk 3]
```diff
@@ -160,7 +160,7 @@ class MontyTypingError(MontyError):
     """Raised when type checking rejects a fed snippet.
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply crates/monty-python/python/pydantic_monty/_monty.pyi:161]: done.

File: crates/monty-js/ts/options.ts [hunk 4]
```diff
@@ -3,6 +3,47 @@
...
+  jsonlines: 5,
+  rdjson: 6,
+  pylint: 7,
+  gitlab: 8,
+  github: 9,
+}
+
+/** Encodes a {@link TypeCheckFormat} to its wire number, throwing on an unknown name. */
+export function encodeTypeCheckFormat(format: TypeCheckFormat): number {
+  const encoded = TYPE_CHECK_FORMATS[format]
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply crates/monty-js/ts/options.ts]: Response from Claude Opus 5 (Claude Code):

Valid — fixed in `encodeTypeCheckFormat`, using the suggested own-property check.

Confirmed the hole is real: `TYPE_CHECK_FORMATS["toString"]` returns `Object.prototype.toString` (a `Function`), so `encoded === undefined` was false and the function returned it as a wire number. `create.uint(7, ...)` then gets a function instead of the `RangeError` a JS caller should have seen. The TypeScript type does not protect this — it is runtime validation precisely because JS callers are not bound by it.

Only the wasm transport was affected; the napi path hands the string to Rust `TypeCheckingFormat::from_name`, a strum match with no prototype chain.

Regression test in `__test__/type_check.spec.ts` covers `toString`, `constructor`, `__proto__`, `hasOwnProperty` and a plain unknown name, and asserts a real name still encodes.

File: crates/monty-proto/src/worker.rs [hunk 5]
```diff
@@ -822,6 +832,7 @@ impl Child {
         self.state = SessionState::Configured(None);
         self.type_check = None;
         self.script_name = String::new();
+        self.type_checker.reset().unwrap();
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply crates/monty-proto/src/worker.rs]: Response from Claude Opus 5 (Claude Code):

Agreed that the `unwrap()` is wrong, but fixed differently: a `FatalError` that stops the child, not a recoverable `protocol_violation`.

`TypeChecker::reset` scrubs the finished session's files out of the type-check database, and it has already `mem::take`n `touched_files` by the time it can fail. So on failure the checker still holds that session's modules and stubs with nothing left tracking them — keeping the worker in service would let the *next* session resolve the previous session's code, which is the security invariant the reset exists for. A session-preserving error would hand that worker straight back to the pool.

`FatalError` gets your actual point — the parent sees a logged explanation instead of an unexplained crash — without returning a contaminated worker to service. It follows the existing version-skew precedent in `Child::handle`:

```rust
pb::parent_request::Kind::Reset(_) => match self.reset() {
    Ok(()) => ok_event(),
    Err(err) => {
        sink.send(&self.fatal_event(&format!("type-check cleanup failed: {err}")))?;
        return Ok(HandleOutcome::Fatal);
    }
},
```

`Child::reset` now returns `Result<(), String>` rather than unwrapping internally.

No test for the error arm: the failure requires the in-memory filesystem to fail a removal on a path it just wrote, which is unreachable without fault injection. The `Ok` path stays covered by the existing pool/subprocess reset tests.

File: crates/monty-runtime/src/main.rs [hunk 6]
```diff
@@ -401,22 +419,30 @@ fn run_repl(file_path: &str, code: &str, tracker: ResourceTracker, mut mount_tab
...
             return ExitCode::FAILURE;
         }
     };
 
     let mut pending_snippet = String::new();
     let mut continuation_mode = ReplContinuationMode::Complete;
 
+    // rustyline writes the prompt itself, so `anstream` never sees it — style
+    // it up front, or not at all when styling is off.
+    let statement_prompt = if stderr_styled() {
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply crates/monty-runtime/src/main.rs]: Response from Claude Opus 5 (Claude Code):

Valid — fixed, prompt styling now comes from stdout via a `stdout_styled()` helper.

Confirmed rustyline writes the prompt to stdout on both its paths: `STDOUT_FILENO` in the normal editing path (`tty/unix.rs`), and an explicit `io::stdout()` in the unsupported-terminal path (`lib.rs:656`).

Reproduced the leak with `monty -i > out.txt` from an interactive terminal (stdin + stderr on ptys, stdout to a file):

```
before: b"\x1b[?2004h\r\x1b[K\x1b[1m\x1b[36m\xe2\x9d\xaf\x1b[0m \r\x1b[2Cexit..."
after:  b"\x1b[?2004h\r\x1b[K\xe2\x9d\xaf \r\x1b[2Cexit..."
```

The remaining `\x1b[?2004h` is rustyline's own bracketed-paste sequence, which is its business rather than ours.

`stderr_styled()` stays for the type checker's `color` flag, since those diagnostics go out through `eprintln!` — the two now ask about the stream each one actually writes to.

No automated test: reproducing it needs a pty on stderr with stdout redirected, which is unix-only and outside what the cross-platform test suites here can express (per CLAUDE.md). Verified manually both ways — redirected stdout is unstyled, a terminal stdout still gets the cyan prompt.
